### ISSUE TITLE
gRPC library is not deinitialized after stream read operations

### ISSUE THREAD

**Issue by nikola-sh** [grpc/grpc#31045]
### What version of gRPC and what language are you using?
https://github.com/grpc/grpc/commit/9d047f88f5762b9353b8f2b5b161f37ff6134fe1
C++

### What operating system (Linux, Windows,...) and version?
Windows 10
Linux

### What runtime / compiler are you using (e.g. python version or version of gcc)
VS 2022
gcc 11.2.0

### What did you do?
I noticed that my gRPC client application crashes sometimes after `main`.
It was happened because gRPC doesn't deinitialize itself after stream operations even if all gRPC objects are destroyed.
To reproduce this situation I added to the `helloworld` sample async streaming operations and gRPC deinitialization waiting at the end.
https://github.com/nikola-sh/grpc/commit/3c7d8bcf281b18d277efff3eb5cfd5710c9d4a4a

If we run the server that sends 1 message and client
```
> greeter_async_server.exe 1
> greeter_async_client2.exe
```
client reads message and is waiting in `while (grpc_is_initialized())` forever
```
started
ok: Hello 123
!ok
Waiting GRPC shutdown
Waiting GRPC shutdown
Waiting GRPC shutdown
Waiting GRPC shutdown
...
```

But if you run server that sends 0 message all will be fine.
```
> greeter_async_server.exe 0
> greeter_async_client2.exe
```
client output:
```
started
!ok
```

In case of 10 messages problem reproduction is unstable, sometimes we see `Waiting GRPC shutdown`, but sometimes not.

### What did you expect to see?
I know that gRPC deinitialization can be asynchronous but I expect that waiting `while (grpc_is_initialized())` is finite and ends fast.

**Comment by yashykt** [comment]
`grpc_is_initialized` is not part of the public API for gRPC C++. It is a gRPC Core API which is only meant to be used internally by the gRPC wrapped languages. At this point, it is considered something akin to an implementation detail.

**Comment by nikola-sh** [comment]
@yashykt , grpc_is_initialized was used only to highlight the problem.
The problem is

> I noticed that my gRPC client application crashes sometimes after main.
> It was happened because gRPC doesn't deinitialize itself after stream operations even if all gRPC objects are destroyed.

C/C++ runtime deinitializes itself after main and alive GRPC threads crash.

**Comment by nikola-sh** [comment]
The problem is fixed in v1.51.0.
Probably by https://github.com/grpc/grpc/pull/30992
